### PR TITLE
fix(versioning/hashicorp): allow `!=` in ranges, constraints

### LINKED ISSUE #22417: feat(versioning/hashicorp): support version excludes `!=`
### What would you like Renovate to be able to do?

We currently do not support exclusions ( `!=` ) as these have no equivalent in NPM to which we currently translate ranges. 

To support the full [syntax](https://developer.hashicorp.com/terraform/language/expressions/version-constraints#version-constraint-syntax) this functionality is needed. 

### If you have any ideas on how this should be implemented, please tell us here.

To implement this functionality we have to remove exclusions before converting and than removing the matching versions. 

Context: https://github.com/renovatebot/renovate/discussions/22414 

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
remove exclusions before converting and than removing the matching versions.

## Context
Closes https://github.com/renovatebot/renovate/issues/22417

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
Test Run : https://github.com/RahulGautamSingh-testing/renovate-tf-lockfile-bug

File: lib/modules/versioning/hashicorp/index.spec.ts [hunk 1]
```diff
@@ -13,9 +14,10 @@ describe('modules/versioning/hashicorp/index', () => {
   );
 
   it.each`
-    versions                                         | range         | expected
-    ${['0.4.0', '0.5.0', '4.0.0', '4.2.0', '5.0.0']} | ${'~> 4.0'}   | ${'4.2.0'}
-    ${['0.4.0', '0.5.0', '4.0.0', '4.2.0', '5.0.0']} | ${'~> 4.0.0'} | ${'4.0.0'}
+    versions                                         | range                 | expected
+    ${['0.4.0', '0.5.0', '4.0.0', '4.2.0', '5.0.0']} | ${'~> 4.0'}           | ${'4.2.0'}
+    ${['0.4.0', '0.5.0', '4.0.0', '4.2.0', '5.0.0']} | ${'~> 4.0.0'}         | ${'4.0.0'}
+    ${['0.4.0', '0.5.0', '4.0.0', '4.2.0', '5.0.0']} | ${'!=4.2.0, > 4.0.0'} | ${'5.0.0'}
```
Comment (secustor) [inline lib/modules/versioning/hashicorp/index.spec.ts:20]:
```suggestion
    ${['0.4.0', '0.5.0', '4.1.0', '4.2.0', '5.0.0']} | ${'!=4.2.0, > 4.0.0'} | ${'4.1.0'}
```

Comment (RahulGautamSingh) [reply lib/modules/versioning/hashicorp/index.spec.ts:20]: getSatisfyingVersion returns highest among the supplied array, so it will be `5.0.0` right?

Comment (secustor) [reply lib/modules/versioning/hashicorp/index.spec.ts:21]: Ah, I have misread that. 

Add my edit as additional test

File: lib/modules/versioning/common.ts [hunk 2]
```diff
@@ -5,3 +5,17 @@ export function isVersioningApiConstructor(
 ): obj is VersioningApiConstructor {
   return typeof obj === 'function';
 }
+
+export function getExcludedVersions(range: string): string[] {
+  return range
+    .split(',')
+    .filter((version) => version.match('!='))
```
Comment (viceice) [inline lib/modules/versioning/common.ts:12]:
```suggestion
    .map(v => v.trim())
    .filter((version) => /^!=/.test(version))
```
this is much faster

File: lib/modules/versioning/common.ts [hunk 3]
```diff
@@ -5,3 +5,17 @@ export function isVersioningApiConstructor(
...
+    .map((version) => version.replace('!=', '').trim());
+}
+
+export function getFilteredRange(range: string): string {
+  return range
+    .split(',')
+    .filter((version) => !version.match('!='))
```
Comment (viceice) [inline lib/modules/versioning/common.ts:19]: same as above

**Review by viceice — APPROVED** [review]

**Pull request merged**